tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/TypeUnifier.scala

```diff
@@ -0,0 +1,311 @@
+package at.forsyte.apalache.tla.typecheck.etc
+
+import at.forsyte.apalache.tla.typecheck._
+import at.forsyte.apalache.tla.typecheck.etc.TypeUnifier.CycleDetected
+
+import scala.collection.immutable.SortedMap
+
+/**
+  * <p>A unification solver for the TlaType1 system. Note that our subtyping relation unifies records
+  * and sparse tuples with a different number of fields. It does so by extending the key set, not by shrinking it.</p>
+  *
+  * <p>This class is not designed for concurrency. Use different instances in different threads.</p>
+  *
+  * @author Igor Konnov
+  */
+class TypeUnifier {
+  // a partial solution to the unification problem is stored here during unification
+  private var solution: Map[Int, TlaType1] = Map.empty
+
+  def unify(substitution: Substitution, lhs: TlaType1, rhs: TlaType1): Option[(Substitution, TlaType1)] = {
+    // start with the substitution
+    solution = substitution.context
+    // try to unify
+    try {
+      compute(lhs, rhs) match {
+        case None => // no unifier
+          None
+
+        case Some(unifiedType) =>
+          val result =
+            if (isCyclic) {
+              None
+            } else {
+              computeClosureWhenAcyclic()
+              val substitution = new Substitution(solution)
+              Some((substitution, substitution(unifiedType)))
+            }
+
+          solution = Map.empty // let GC collect the solution map later
+          result
+      }
+    } catch {
+      case _: CycleDetected =>
+        solution = Map.empty // let GC collect the solution map later
+        None
+    }
+  }
+
+  private def computeOptions(lhs: Option[TlaType1], rhs: Option[TlaType1]): Option[TlaType1] = {
```

Comment:
This function is deprecated in the following commits. So let's keep it like that and then remove in the end.

Reply:
I don't see any sign of the function being deprecated on your initial prototyping branch: https://github.com/informalsystems/apalache/pull/244/files#diff-12379bb1bd9390b219a628ad755afb666f5fcb18f250fc21dc66d094353f84b2R68-R74

Reply:
Oh, sorry. I thought of another version of `unify`. Yeah, `computeOptions` is a bit weird.

Reply:
No worries! There was a deprecated `unify` method in here, but I already removed it. :)

> computeOptions is a bit weird

What do you think of my suggestion to just rework it to 

```scala
def computeRecordField(field: string, lhsRecord: RecT1, rhsRecord: Rect1): Option[TlaType1] ...
```

Reply:
I think you are right. It was probably used in another piece of code that is gone by now. Please go ahead and rework it.